File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/play_clues.go
package main

import (
	"encoding/json"
	"net/http"
)

// playClue is a piece of campaign lore the DM may reveal to one character,
// the whole party, or nobody (hidden, for DM record-keeping only).
type playClue struct {
	ClueID      string
	Text        string
	Audience    string
	CharacterID string
}

func playClueResponse(clue *playClue) map[string]interface{} {
	resp := map[string]interface{}{
		"clue_id":  clue.ClueID,
		"text":     clue.Text,
		"audience": clue.Audience,
	}
	if clue.Audience == "character" {
		resp["character_id"] = clue.CharacterID
	}
	return resp
}

// handlePlayCampaignCluesSub routes the "clues" sub-path of a play campaign.
// It returns false if rest does not name the clues path, so the caller can
// fall through to its own routing.
func handlePlayCampaignCluesSub(w http.ResponseWriter, r *http.Request, campaignID, rest string) bool {
	if rest != "clues" {
		return false
	}
	switch r.Method {
	case http.MethodPost:
		handleCreatePlayClue(w, r, campaignID)
	case http.MethodGet:
		handleListPlayClues(w, r, campaignID)
	default:
		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
	}
	return true
}

// handleCreatePlayClue lets the campaign dm create a new clue targeted to a
// character, the party, or hidden.
func handleCreatePlayClue(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req struct {
		ClueID      string `json:"clue_id"`
		Text        string `json:"text"`
		Audience    string `json:"audience"`
		CharacterID string `json:"character_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.ClueID == "" || req.Text == "" {
		writeError(w, http.StatusBadRequest, "clue_id and text are required")
		return
	}
	switch req.Audience {
	case "character":
		if req.CharacterID == "" {
			writeError(w, http.StatusBadRequest, "character_id is required for character audience")
			return
		}
	case "party", "hidden":
		if req.CharacterID != "" {
			writeError(w, http.StatusBadRequest, "character_id must be omitted for this audience")
			return
		}
	default:
		writeError(w, http.StatusBadRequest, "audience must be character, party, or hidden")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the dm may create clues")
		return
	}
	if req.Audience == "character" && findPlayMemberByCharacterID(c, req.CharacterID) == nil {
		playMu.Unlock()
		writeError(w, http.StatusBadRequest, "unknown character")
		return
	}
	for _, clue := range c.Clues {
		if clue.ClueID == req.ClueID {
			playMu.Unlock()
			writeError(w, http.StatusConflict, "clue_id already exists")
			return
		}
	}

	clue := &playClue{
		ClueID:      req.ClueID,
		Text:        req.Text,
		Audience:    req.Audience,
		CharacterID: req.CharacterID,
	}
	c.Clues = append(c.Clues, clue)
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusCreated, playClueResponse(clue))
}

// handleListPlayClues returns clues visible to the requesting campaign
// member: the dm sees all clues in insertion order, a player sees party
// clues and clues targeted to their own character.
func handleListPlayClues(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if !isPlayMember(c, username) {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the owner or a member may view clues")
		return
	}

	isDM := c.Owner == username
	var ownCharacterID string
	if !isDM {
		for _, m := range c.Members {
			if m.Username == username {
				ownCharacterID = m.CharacterID
				break
			}
		}
	}

	clues := make([]map[string]interface{}, 0, len(c.Clues))
	for _, clue := range c.Clues {
		if isDM {
			clues = append(clues, playClueResponse(clue))
			continue
		}
		switch clue.Audience {
		case "party":
			clues = append(clues, playClueResponse(clue))
		case "character":
			if clue.CharacterID == ownCharacterID {
				clues = append(clues, playClueResponse(clue))
			}
		}
	}
	playMu.Unlock()

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"clues": clues,
	})
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_clues.php
<?php
declare(strict_types=1);

// Routes: Play (campaign-scoped DM clues revealed to a character, the party, or nobody)
// ---------------------------------------------------------------------------

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/clues$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if ($actor['username'] !== $campaign['owner']) {
        forbidden('only the campaign dm may create clues');
    }

    $body = read_json_body();
    if ($body === null
        || !isset($body['clue_id'], $body['text'], $body['audience'])
        || !is_string($body['clue_id']) || $body['clue_id'] === ''
        || !is_string($body['text']) || $body['text'] === ''
        || !is_string($body['audience'])) {
        bad_request();
    }
    $clueId = $body['clue_id'];
    $text = $body['text'];
    $audience = $body['audience'];

    if (!in_array($audience, ['character', 'party', 'hidden'], true)) {
        bad_request('invalid audience');
    }

    if ($audience === 'character') {
        if (!isset($body['character_id']) || !is_string($body['character_id']) || $body['character_id'] === '') {
            bad_request('character_id is required for character audience');
        }
        $characterId = $body['character_id'];
        if (find_play_character($db, $campaignId, $characterId) === null) {
            bad_request('character_id must name a campaign member character');
        }
    } else {
        if (isset($body['character_id'])) {
            bad_request('character_id must be omitted for this audience');
        }
        $characterId = null;
    }

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_clues WHERE campaign_id = ? AND clue_id = ?');
    $stmt->execute([$campaignId, $clueId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('clue id already exists in this campaign');
    }

    $clue = [
        'clue_id' => $clueId,
        'text' => $text,
        'audience' => $audience,
    ];
    if ($audience === 'character') {
        $clue['character_id'] = $characterId;
    }

    $stmt = $db->prepare('SELECT COALESCE(MAX(seq), 0) + 1 FROM play_campaign_clues WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    $seq = (int)$stmt->fetchColumn();

    $stmt = $db->prepare('INSERT INTO play_campaign_clues (campaign_id, clue_id, seq, data) VALUES (?, ?, ?, ?)');
    $stmt->execute([$campaignId, $clueId, $seq, json_encode($clue)]);

    send_json($clue, 201);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/clues$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $isDm = $actor['username'] === $campaign['owner'];
    if (!$isDm && !is_play_campaign_member($db, $campaignId, $actor['username'])) {
        forbidden('only the owner or a party member may view clues');
    }

    $stmt = $db->prepare('SELECT data FROM play_campaign_clues WHERE campaign_id = ? ORDER BY seq ASC');
    $stmt->execute([$campaignId]);
    $rows = $stmt->fetchAll(PDO::FETCH_COLUMN);
    $clues = array_map(fn($row) => json_decode($row, true), $rows);

    if (!$isDm) {
        $stmt = $db->prepare('SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?');
        $stmt->execute([$campaignId, $actor['username']]);
        $ownCharacterId = $stmt->fetchColumn();

        $clues = array_values(array_filter($clues, function ($clue) use ($ownCharacterId) {
            if ($clue['audience'] === 'party') {
                return true;
            }
            if ($clue['audience'] === 'character') {
                return $clue['character_id'] === $ownCharacterId;
            }
            return false;
        }));
    }

    send_json(['clues' => $clues], 200);
}
